test/integration/compound/compound_test.go

```diff
@@ -0,0 +1,426 @@
+/*
+ * // SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+ * //
+ * // SPDX-License-Identifier: Apache-2.0
+ */
+
+package compound_test
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/gardener/controller-manager-library/pkg/controllermanager"
+	"github.com/gardener/external-dns-management/pkg/dns"
+	"github.com/gardener/external-dns-management/pkg/dns/provider"
+	. "github.com/gardener/gardener/pkg/utils/test/matchers"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/controller-manager-library/pkg/ctxutil"
+	"github.com/gardener/external-dns-management/pkg/apis/dns/v1alpha1"
+	"github.com/gardener/external-dns-management/pkg/controller/provider/mock"
+)
+
+var _ = Describe("Compound controller tests", func() {
+	const (
+		defaultTTL   = 300
+		retryTimeout = 5 * time.Second
+	)
+
+	var (
+		testRunID      string
+		testNamespace  *corev1.Namespace
+		firstZoneID    dns.ZoneID
+		e1, e2, e3, e4 *v1alpha1.DNSEntry
+
+		checkDeleted = func(g Gomega, ctx context.Context, entry *v1alpha1.DNSEntry) {
+			err := testClient.Get(ctx, client.ObjectKeyFromObject(entry), entry)
+			g.Expect(err).To(HaveOccurred())
+			g.Expect(client.IgnoreNotFound(err)).To(Succeed())
+		}
+
+		checkSingleEntryInMockDatabaseGomega = func(g Gomega, entry *v1alpha1.DNSEntry) {
+			dump := mock.TestMock[testRunID].BuildFullDump()
+			for _, zoneDump := range dump.InMemory {
+				switch {
+				case zoneDump.HostedZone.Domain == "first.example.com" && entry == nil:
+					g.Expect(zoneDump.DNSSets).To(BeEmpty(), "unexpected number of DNS sets in first.example.com")
+				case zoneDump.HostedZone.Domain == "first.example.com" && entry != nil:
+					g.Expect(zoneDump.DNSSets).To(HaveKey(dns.DNSSetName{DNSName: entry.Spec.DNSName}))
+					g.Expect(zoneDump.DNSSets).To(HaveLen(1), "unexpected number of DNS sets in first.example.com")
+					set := zoneDump.DNSSets[dns.DNSSetName{DNSName: entry.Spec.DNSName}]
+					g.Expect(set.Sets).To(HaveKey("A"))
+					g.Expect(set.Sets).To(HaveKey("META"))
+					g.Expect(set.Sets).To(HaveLen(2))
+					setA := set.Sets["A"]
+					g.Expect(setA.Records).To(ConsistOf(
+						PointTo(MatchFields(IgnoreExtras, Fields{
+							"Value": Equal(entry.Spec.Targets[0]),
+						}))))
+					g.Expect(setA.Type).To(Equal("A"))
+					g.Expect(setA.TTL).To(Equal(int64(defaultTTL)))
+				case zoneDump.HostedZone.Domain == "second.example.com":
+					g.Expect(zoneDump.DNSSets).To(BeEmpty(), "unexpected number of DNS sets in second.example.com")
+				default:
+					g.Expect(false).To(BeTrue(), "unexpected zone domain "+zoneDump.HostedZone.Domain)
```

Comment:
While mostly a semantic difference I think using `Fail()` might be more appropriate here
```suggestion
					Fail("unexpected zone domain "+zoneDump.HostedZone.Domain))
```